rectools/models/candidate_ranking.py

```diff
@@ -0,0 +1,637 @@
+            k=self.num_candidates,
+            filter_viewed=filter_viewed,
+            items_to_recommend=items_to_recommend,
+            add_rank_col=self.keep_ranks,
+            on_unsupported_targets=on_unsupported_targets,
+        )
+        if not self.keep_scores:
+            candidates.drop(columns=Columns.Score, inplace=True)
+        return candidates
+
+
+class CandidateRankingModel(ModelBase):
+    """
+    Candidate Ranking Model for recommendation systems.
+    """
+
+    def __init__(
+        self,
+        candidate_generators: tp.List[CandidateGenerator],
+        splitter: Splitter,
+        reranker: Reranker = CatBoostReranker(),
+        sampler: NegativeSamplerBase = PerUserNegativeSampler(),
+        feature_collector: CandidateFeatureCollector = CandidateFeatureCollector(),
+        verbose: int = 0,
+    ) -> None:
+        """
+        Initialize the CandidateRankingModel with candidate generators, splitter, reranker, sampler
+        and feature collector.
+
+        Parameters
+        ----------
+        candidate_generators : tp.List[CandidateGenerator]
+            List of candidate generators.
+        splitter : Splitter
+            Splitter for dataset splitting.
+        reranker : Reranker
+            Reranker for reranking candidates.
+        sampler : NegativeSamplerBase, optional
+            Sampler for negative sampling. Default is PerUserNegativeSampler().
+        feature_collector : CandidateFeatureCollector, optional
+            Collector for user-item features. Default is CandidateFeatureCollector().
+        verbose : int, optional
+            Verbosity level. Default is 0.
+        """
+
+        super().__init__(verbose=verbose)
+
+        if hasattr(splitter, "n_splits"):
+            assert splitter.n_splits == 1  # TODO: handle softly
+        self.splitter = splitter
+        self.sampler = sampler
+        self.reranker = reranker
+        self.cand_gen_dict = self._create_cand_gen_dict(candidate_generators)
+        self.feature_collector = feature_collector
+
+    def _create_cand_gen_dict(
+        self, candidate_generators: tp.List[CandidateGenerator]
+    ) -> tp.Dict[str, CandidateGenerator]:
+        """
+        Create a dictionary of candidate generators with unique identifiers.
+
+        Parameters
+        ----------
+        candidate_generators : tp.List[CandidateGenerator]
+            List of candidate generators.
+
+        Returns
+        -------
+        tp.Dict[str, CandidateGenerator]
+            Dictionary with candidate generator identifiers as keys and candidate generators as values.
+        """
+        model_count = {}
+        cand_gen_dict = {}
+        for candgen in candidate_generators:
+            model_name = candgen.model.__class__.__name__
+            if model_name not in model_count:
+                model_count[model_name] = 0
+            model_count[model_name] += 1
+            identifier = f"{model_name}_{model_count[model_name]}"
```

Comment:
I don't yet know how exactly do you plan to use identifiers, but do you think it make sense to generate some sort of hash for models?
Or, if they are used for column names (=features), maybe it's better to ask user to provide the names? In this case it would be easier to explore feature importance e.g.

Reply:
We need a dict for feature names in columns, yes.
But user (me) is too lazy to create a dict when he can just pass a list.
It's just a bit faster. I would prefer to leave it at it is.